src/withdrawals/tasks.py

```diff
@@ -0,0 +1,351 @@
+                val.public_key: val.balance for val in partial_withdrawable_validators
+            },
+            queued_assets=queued_assets,
+        )
+    if settings.disable_full_withdrawals:
+        logger.info('Available partial withdrawals capacity is less than queued assets')
+        return {}
+
+    min_activation_epoch = max(0, chain_head.epoch - validator_min_active_epochs)
+
+    can_be_exited_validators = _filter_exitable_validators(
+        consensus_validators=consensus_validators,
+        min_activation_epoch=min_activation_epoch,
+        oracle_exit_indexes=oracle_exit_indexes,
+    )
+    withdrawals_data: dict[HexStr, Gwei] = {}
+
+    for validator in can_be_exited_validators:
+        if sum(withdrawals_data.values()) >= queued_assets:
+            return withdrawals_data
+        if sum(withdrawals_data.values()) + available_partial_withdrawals_capacity >= queued_assets:
+            partial_data = _get_partial_withdrawals_data(
+                validator_balances={
+                    val.public_key: val.balance for val in partial_withdrawable_validators
+                },
+                queued_assets=queued_assets,
+            )
+            return withdrawals_data | partial_data
+
+        partial_withdrawable_validators = [
+            val for val in partial_withdrawable_validators if val.public_key != validator.public_key
+        ]
+        queued_assets = Gwei(queued_assets - validator.balance)
+        available_partial_withdrawals_capacity = sum(
+            val.balance - MIN_ACTIVATION_BALANCE_GWEI for val in partial_withdrawable_validators
+        )
+
+        withdrawals_data[validator.public_key] = Gwei(0)  # full withdrawal
+
+    return withdrawals_data
+
+
+def _get_partial_withdrawals_data(
+    validator_balances: dict[HexStr, Gwei], queued_assets: Gwei
+) -> dict[HexStr, Gwei]:
+    """
+    Calculate partial withdrawals for validators.
+    First, try to withdraw from a single validator in one request.
+    If that fails, split the amount across validators,
+    starting with those holding the largest balances.
+    """
+    withdrawals_data: dict[HexStr, Gwei] = {}
+
+    # Try single validator first
+    for public_key, balance in sorted(
+        validator_balances.items(), key=lambda item: item[1], reverse=True
+    ):
+        available = balance - MIN_ACTIVATION_BALANCE_GWEI
+        if available >= queued_assets:
+            return {public_key: queued_assets}
+
+    # Split across multiple validators
+    for public_key, balance in sorted(
+        validator_balances.items(), key=lambda item: item[1], reverse=True
+    ):
+        available = balance - MIN_ACTIVATION_BALANCE_GWEI
+        if available <= 0:
+            continue
+
+        amount = Gwei(min(available, queued_assets))
+        withdrawals_data[public_key] = amount
+        queued_assets = Gwei(queued_assets - amount)
+
+        if queued_assets <= 0:
+            break
+
+    return withdrawals_data
```

Comment:
Don't get why we can't process it in single iteration, starting from the validator with the highest balance and until we run out of queued assets.